### PR TITLE
Add Zones field to InstanceGroup

### PR DESCRIPTION
**Pull request by justinsb**
The Zones field can specify zones where they are not specified on a
Subnet, for example on GCE where we have regional subnets.

### CODE REVIEW

**Comment by justinsb** [comment]
#3438 should have the first 3 commits, so this is only the last commit.  Will rebase after #3438.

Split because otherwise the PR is harder to read (though you can always go commit by commit), but I'm not sure whether I'll be able to get #3438 to pass without this PR.  Fingers crossed :-)

File: cmd/kops/create_cluster.go [hunk 1]
```diff
@@ -387,29 +389,70 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
...
-			existingSubnets[subnet.Name] = subnet
+	if c.Cloud != "" {
+		cluster.Spec.CloudProvider = c.Cloud
+	}
+
+	allZones := sets.NewString()
+	allZones.Insert(c.Zones...)
+	allZones.Insert(c.MasterZones...)
+
+	if cluster.Spec.CloudProvider == "" {
```
Comment (justinsb) [inline cmd/kops/create_cluster.go]: This is just moved up from lower...

File: cmd/kops/create_cluster.go [hunk 2]
```diff
@@ -387,29 +389,70 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
...
 					Egress: c.Egress,
-				})
+				}
+				cluster.Spec.Subnets = append(cluster.Spec.Subnets, *subnet)
 			}
+			zoneToSubnetMap[zoneName] = subnet
 		}
 	}
 
-	if len(cluster.Spec.Subnets) == 0 {
```
Comment (justinsb) [inline cmd/kops/create_cluster.go]: This is covered by the explicit Zones check above

File: cmd/kops/create_cluster.go [hunk 3]
```diff
@@ -424,55 +467,55 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
...
-				if subnet == nil {
-					// Should have been caught already
-					return fmt.Errorf("master-zone %q not included in zones", subnetName)
-				}
-				masterSubnets = append(masterSubnets, subnet)
-			}
-
-			if c.MasterCount != 0 && c.MasterCount < int32(len(masterSubnets)) {
-				return fmt.Errorf("specified %d master zones, but also requested %d masters.  If specifying both, the count should match.", len(masterSubnets), c.MasterCount)
+		masterZones := c.MasterZones
```
Comment (justinsb) [inline cmd/kops/create_cluster.go]: I think this logic is clearer and fixes the bug when you specify zones and master-count

File: cmd/kops/create_cluster.go [hunk 4]
```diff
@@ -535,8 +574,22 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
 	if len(nodes) == 0 {
 		g := &api.InstanceGroup{}
 		g.Spec.Role = api.InstanceGroupRoleNode
-
 		g.ObjectMeta.Name = "nodes"
+
+		subnetNames := sets.NewString()
```
Comment (justinsb) [inline cmd/kops/create_cluster.go]: We explicitly initialize it - it is safer, and we don't necessarily have all the information later in the GCE case

File: pkg/apis/kops/v1alpha1/conversion.go [hunk 5]
```diff
@@ -235,15 +235,25 @@ func Convert_kops_EtcdMemberSpec_To_v1alpha1_EtcdMemberSpec(in *kops.EtcdMemberS
 }
 
 func Convert_v1alpha1_InstanceGroupSpec_To_kops_InstanceGroupSpec(in *InstanceGroupSpec, out *kops.InstanceGroupSpec, s conversion.Scope) error {
+	err := autoConvert_v1alpha1_InstanceGroupSpec_To_kops_InstanceGroupSpec(in, out, s)
```
Comment (justinsb) [inline pkg/apis/kops/v1alpha1/conversion.go]: Changed order here to preserve the logic, now that the autogenerated code has Zones fields...

File: pkg/apis/kops/validation/legacy.go [hunk 6]
```diff
@@ -76,14 +76,24 @@ func ValidateCluster(c *kops.Cluster, strict bool) *field.Error {
 
 	requiresSubnets := true
 	requiresNetworkCIDR := true
+	requiresSubnetCIDR := true
 	switch kops.CloudProviderID(c.Spec.CloudProvider) {
 	case kops.CloudProviderBareMetal:
 		requiresSubnets = false
 		requiresNetworkCIDR = false
+		if c.Spec.NetworkCIDR != "" {
+			return field.Invalid(fieldSpec.Child("NetworkCIDR"), c.Spec.NetworkCIDR, "NetworkCIDR should not be set on bare metal")
```
Comment (justinsb) [inline pkg/apis/kops/validation/legacy.go]: As the NetworkCIDR is ignored, makes more sense to error if it is set

File: upup/pkg/fi/cloudup/defaults.go [hunk 7]
```diff
@@ -43,29 +43,31 @@ func PerformAssignments(c *kops.Cluster) error {
 		return err
 	}
 
-	if c.SharedVPC() && c.Spec.NetworkCIDR == "" {
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/defaults.go]: Same code just refactored, and only done for AWS now

File: upup/pkg/fi/cloudup/gce/utils.go [hunk 8]
```diff
@@ -69,3 +70,13 @@ func LastComponent(s string) string {
 	}
 	return s
 }
+
+// ZoneToRegion maps a GCE zone name to a GCE region name, returning an error if it cannot be mapped
+func ZoneToRegion(zone string) (string, error) {
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/gce/utils.go]: Pulled this into a helper function

File: upup/pkg/fi/cloudup/utils.go [hunk 9]
```diff
@@ -40,20 +40,13 @@ func BuildCloud(cluster *kops.Cluster) (fi.Cloud, error) {
 	switch kops.CloudProviderID(cluster.Spec.CloudProvider) {
 	case kops.CloudProviderGCE:
 		{
-			nodeZones := make(map[string]bool)
 			for _, subnet := range cluster.Spec.Subnets {
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/utils.go]: This got much easier :-)

**Comment by justinsb** [comment]
#3438 can't pass without this (which is why I had to do it :-) ).  #3438 is the first commit here; not sure whether it's easier to review the two commits here therefore...

File: cmd/kops/create_cluster.go [hunk 10]
```diff
@@ -387,29 +389,70 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
...
+			return fmt.Errorf("unable to infer CloudProvider from Zones (is there a typo in --zones?)")
 		}
-		for _, zoneName := range c.Zones {
+	}
+
+	zoneToSubnetMap := make(map[string]*api.ClusterSubnetSpec)
+	if len(c.Zones) == 0 {
+		return fmt.Errorf("must specify at least one zone for the cluster (use --zones)")
+	} else if api.CloudProviderID(cluster.Spec.CloudProvider) == api.CloudProviderGCE {
+		// On GCE, subnets are regional - we create one per region, not per zone
```
Comment (chrislovecnm) [inline cmd/kops/create_cluster.go]: Should we have this code in the cloud provider rather than create?  Move it into an interfaced method as all cloud providers would have to implement it.

File: pkg/apis/kops/cluster.go [hunk 11]
```diff
@@ -324,8 +325,10 @@ type ClusterSubnetSpec struct {
 	Name string `json:"name,omitempty"`
 	// CIDR is the network cidr of the subnet
 	CIDR string `json:"cidr,omitempty"`
-	// Zone is the zone the subnet resides
+	// Zone is the zone the subnet is in, set for subnets that are zonally scoped
```
Comment (chrislovecnm) [inline pkg/apis/kops/cluster.go]: Can we just have one value to use across all providers?

File: pkg/apis/kops/v1alpha1/conversion.go (hunk 5 above)
Comment (chrislovecnm) [reply pkg/apis/kops/v1alpha1/conversion.go]: Will this impact auto-generated code?

File: pkg/apis/kops/validation/legacy.go (hunk 6 above)
Comment (chrislovecnm) [reply pkg/apis/kops/validation/legacy.go]: Can we move this to another PR?  I love it, but it kinda does not apply to the work you are doing.

File: pkg/apis/kops/validation/legacy.go [hunk 12]
```diff
@@ -76,14 +76,24 @@ func ValidateCluster(c *kops.Cluster, strict bool) *field.Error {
...
+		}
+
+	case kops.CloudProviderGCE:
```
Comment (chrislovecnm) [inline pkg/apis/kops/validation/legacy.go]: Same comment as above.  Can we break this into a different PR?

File: pkg/apis/kops/v1alpha1/conversion.go (hunk 5 above)
Comment (justinsb) [reply pkg/apis/kops/v1alpha1/conversion.go]: Should not.

File: cmd/kops/create_cluster.go (hunk 10 above)
Comment (justinsb) [reply cmd/kops/create_cluster.go]: We can see how this goes I suspect as we add more clouds

File: pkg/apis/kops/cluster.go (hunk 11 above)
Comment (justinsb) [reply pkg/apis/kops/cluster.go]: Not sure I understand

File: pkg/apis/kops/validation/legacy.go (hunk 12 above)
Comment (justinsb) [reply pkg/apis/kops/validation/legacy.go]: Will move the bare metal one into another PR, but it's not really practical to split out the GCE logic here..

Comment (justinsb) [reply pkg/apis/kops/validation/legacy.go]: Done in #3495

**Comment by justinsb** [comment]
Trivial rebase (imports), will reapply LGTM
